src/Identity/IdentityServer/BaseRequestValidator.cs

```diff
@@ -133,6 +142,15 @@ public abstract class BaseRequestValidator<T> where T : class
             var verified = await VerifyTwoFactor(user, twoFactorOrganization,
                 twoFactorProviderType, twoFactorToken);
 
+            var isOtpCached = Core.Utilities.DistributedCacheExtensions.TryGetValue(_distributedCache, user.Email, out string cachedToken);
+            if (isOtpCached)
+            {
+                // Delay for brute force.
+                await Task.Delay(2000);
+                await BuildErrorResultAsync("Two-step token has already been used.", true, context, user);
```

Comment:
💭 Should we reveal this vs. just saying it's invalid?

Reply:
Yes

Reply:
Error message updated.

Reply:
Samesies.